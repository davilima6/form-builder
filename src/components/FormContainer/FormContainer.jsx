// @flow
import React, { useCallback, lazy, Suspense } from 'react';
import { withRouter } from 'react-router-dom';
import styled from 'styled-components';
import CONFIG from '../../config';
import {
  Error,
  Field,
  hasErrors,
  getMinAgeMinDateFrom,
  normalize,
  useScrollToTop,
  useForm,
} from '../../utils';
import {
  errorMessages,
  validateInRange,
  validateMinDate,
  validateMaxLength,
  validatePattern,
  validateRequired,
} from '../../utils/validators';
import { Spinner } from '..';

const Form = lazy(() => import('../Form/Form'));

function validate(schema: { [string]: Field }, data: { [string]: Error }): Array<Error> {
  const errors = schema.reduce((obj, field) => {
    const fieldName = normalize(field.label);
    const currentValue = data[fieldName] || '';

    // Validate all fields as required
    let fieldErrors = [...validateRequired(currentValue)];

    switch (field.type) {
      case 'email':
        fieldErrors = [
          ...fieldErrors,
          ...validateMaxLength(currentValue, CONFIG.rules.email.max),
          ...validatePattern(currentValue, CONFIG.rules.email.pattern, CONFIG.rules.email.domain),
        ];
        break;
      case 'text':
        fieldErrors = [...fieldErrors, ...validateMaxLength(currentValue, CONFIG.rules.text.max)];
        break;
      default:
    }

    switch (fieldName) {
      case 'age':
        fieldErrors = [
          ...fieldErrors,
          ...validateInRange(currentValue, CONFIG.rules.age.min, CONFIG.rules.age.max),
        ];
        break;
      case 'date-of-birth': {
        const currentDate = new Date(currentValue);
        const now = new Date();
        const minDate = getMinAgeMinDateFrom(now);

        fieldErrors = [...fieldErrors, ...validateMinDate(currentDate, minDate)];
        break;
      }
      default:
    }

    // eslint-disable-next-line no-param-reassign
    obj[fieldName] = fieldErrors;

    return obj;
  }, {});

  return errors;
}

function redirectTo(history, toRoute): void {
  history.push(toRoute);
}

const StyledFormContainer = styled.section`
  display: flex;
  justify-content: center;
`;

type Props = {
  history: Array<string>,
};

const FormContainer = ({ history }: Props) => {
  const {
    schema, data, dispatch, errors, setErrors,
  } = useForm();

  useScrollToTop(errors);

  const onSubmit = useCallback(
    (event: SyntheticEvent<>) => {
      event.preventDefault();
      const newErrors = validate(schema, data);
      const hasErrored = hasErrors(newErrors);

      if (!hasErrored) {
        redirectTo(history, 'success');
      }

      setErrors(prevErrors => ({ ...prevErrors, ...newErrors }));
    },
    [data],
  );

  const onCancel = useCallback(() => {
    dispatch({ type: 'reset' });
    setErrors({});
  }, []);

  return (
    <StyledFormContainer>
      <Suspense fallback={<Spinner size={100} />}>
        <Form
          schema={schema}
          data={data}
          dispatch={dispatch}
          errors={errors}
          errorMessages={errorMessages}
          onSubmit={onSubmit}
          onCancel={onCancel}
        />
      </Suspense>
    </StyledFormContainer>
  );
};

export default withRouter(FormContainer);
