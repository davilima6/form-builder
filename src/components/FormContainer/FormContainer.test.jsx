import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { cleanup, render } from 'react-testing-library';
// import 'jest-styled-components';
import 'jest-dom/extend-expect';
import FormContainer from './FormContainer';

window.scrollTo = jest.fn();

describe('Form', () => {
  let component;

  beforeEach(() => {
    cleanup();
    component = render(
      <BrowserRouter>
        <FormContainer />
      </BrowserRouter>,
    );
  });

  test('renders without crashing', () => expect(component).toBeDefined());

  test('matches snapshot', () => {
    const { asFragment } = component;

    expect(asFragment()).toMatchSnapshot();
    // expect(component).toHaveStyleRule('display', 'flex');
  });
});
