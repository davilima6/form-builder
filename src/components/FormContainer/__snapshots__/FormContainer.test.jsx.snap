// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="sc-kgoBCf gnFrFj"
  >
    <form
      class="sc-bwzfXH hJnTjW"
      method="post"
    >
      <div
        class="sc-htpNat BiPJH"
      >
        <h2
          class="sc-bxivhb bUZGRr"
        >
          Title of the form
        </h2>
        <p
          class="sc-ifAKCX fgJOcI"
        >
          Longer description of the form
        </p>
      </div>
      <div
        class="sc-gqjmRU jqyabD"
      >
        <div
          class="sc-fjdhpX fQffyr"
        >
          <label
            class="sc-gZMcBi giNnHm"
            for="field-first-name"
          >
            <div
              class="sc-jTzLTM jXxOTJ"
            >
              First Name
            </div>
            <input
              class="sc-VigVT dQqCzP"
              id="field-first-name"
              name="first-name"
              type="text"
              value=""
            />
          </label>
        </div>
      </div>
      <div
        class="sc-gqjmRU jqyabD"
      >
        <div
          class="sc-fjdhpX fQffyr"
        >
          <label
            class="sc-gZMcBi giNnHm"
            for="field-last-name"
          >
            <div
              class="sc-jTzLTM jXxOTJ"
            >
              Last Name
            </div>
            <input
              class="sc-VigVT dQqCzP"
              id="field-last-name"
              name="last-name"
              type="text"
              value=""
            />
          </label>
        </div>
      </div>
      <div
        class="sc-gqjmRU jqyabD"
      >
        <div
          class="sc-fjdhpX fQffyr"
        >
          <label
            class="sc-gZMcBi giNnHm"
            for="field-email"
          >
            <div
              class="sc-jTzLTM jXxOTJ"
            >
              email
            </div>
            <input
              class="sc-VigVT dQqCzP"
              id="field-email"
              name="email"
              placeholder="your username@google.com"
              type="email"
              value=""
            />
          </label>
        </div>
      </div>
      <div
        class="sc-gqjmRU jqyabD"
      >
        <div
          class="sc-fjdhpX fQffyr"
        >
          <label
            class="sc-gZMcBi giNnHm"
            for="field-age"
          >
            <div
              class="sc-jTzLTM jXxOTJ"
            >
              Age
            </div>
            <input
              class="sc-VigVT dQqCzP"
              id="field-age"
              name="age"
              type="number"
              value=""
            />
          </label>
        </div>
      </div>
      <div
        class="sc-gqjmRU jqyabD"
      >
        <div
          class="sc-iwsKbI kUkwhG sc-fjdhpX fQffyr"
        >
          <div
            class="sc-htoDjs icOsfV sc-jTzLTM jXxOTJ"
          >
            Gender
          </div>
          <label
            class="sc-gzVnrw iVyyZC sc-gZMcBi giNnHm"
            for="field-gender-female"
          >
            <div
              class="sc-dnqmqq kaxzYb"
            >
              Female
            </div>
            <input
              class="sc-VigVT dQqCzP"
              id="field-gender-female"
              name="gender"
              type="radio"
              value="0"
            />
          </label>
          <label
            class="sc-gzVnrw iVyyZC sc-gZMcBi giNnHm"
            for="field-gender-male"
          >
            <div
              class="sc-dnqmqq kaxzYb"
            >
              Male
            </div>
            <input
              class="sc-VigVT dQqCzP"
              id="field-gender-male"
              name="gender"
              type="radio"
              value="1"
            />
          </label>
          <label
            class="sc-gzVnrw iVyyZC sc-gZMcBi giNnHm"
            for="field-gender-non-binary"
          >
            <div
              class="sc-dnqmqq kaxzYb"
            >
              Non-Binary
            </div>
            <input
              class="sc-VigVT dQqCzP"
              id="field-gender-non-binary"
              name="gender"
              type="radio"
              value="2"
            />
          </label>
        </div>
      </div>
      <div
        class="sc-gqjmRU jqyabD"
      >
        <div
          class="sc-fjdhpX fQffyr"
        >
          <label
            class="sc-gZMcBi giNnHm"
            for="field-date-of-birth"
          >
            <div
              class="sc-jTzLTM jXxOTJ"
            >
              Date of Birth
            </div>
            <input
              class="sc-VigVT dQqCzP"
              id="field-date-of-birth"
              name="date-of-birth"
              type="date"
              value=""
            />
          </label>
        </div>
      </div>
      <div
        class="sc-EHOje jFpCHA"
      >
        <button
          class="sc-bdVaJa cdNCer"
          title="Send"
          type="submit"
        >
          Send
        </button>
        <button
          class="sc-bdVaJa gzdvGZ"
          title="Clear"
          type="reset"
        >
          Clear
        </button>
      </div>
    </form>
  </section>
</DocumentFragment>
`;
